draft/0001-organization.md

```diff
@@ -0,0 +1,273 @@
+| organization_roles | array of JSON objects | An array of organization roles that the user has. Each object contains the following fields: <ul><li>`organization_id`: The unique organization identifier as a string.</li><li>`roles`: The role names granted to the user within the organization.</li></ul> |
+
+The objects in the `organization_roles` array must adhere to these rules:
+
+- The `organization_id` field must be unique within the array.
+- The `roles` field must be an array of strings, with each string being unique within the array.
+- The `roles` field must not be empty.
+
+These rules help minimize the size of the ID token and can be enforced by the OpenID Provider (OP) when issuing the ID token.
+
+An example of the ID token with organization claims:
+
+```json
+{
+  "sub": "12345",
+  "name": "John Wick",
+  "organization_ids": ["12345", "67890", "13579"],
+  "organization_roles": [
+    {
+      "organization_id": "12345",
+      "roles": ["admin"]
+    },
+    {
+      "organization_id": "67890",
+      "roles": ["viewer", "editor"]
+    }
+  ]
+}
+```
+
+### 5.2 Requesting claims using scope values
+
+As per the existing authentication flow, the OpenID Provider (OP) SHOULD accept an additional `scope` value in the authentication request and provide the corresponding claim in the ID token:
+
+- `organization`: This scope value requests access to the `organization_ids` and `organization_roles` claims.
+
+## 6. Authorization
+
+We utilize the same Token Endpoint defined in OAuth 2.0 and introduce a new grant type for requesting JWT access tokens in the context of an organization.
+
+Implementing JWT enforcement can enable offline authorization for the resource server, reducing network latency and improving performance.
+
+### 6.1. Token request
+
+OAuth 2.0 defines the [Token Endpoint](https://www.rfc-editor.org/rfc/rfc6749.html#section-3.2) for requesting access tokens. Although it's possible to add more parameters to the existing grants (as [RFC 8707](https://www.rfc-editor.org/rfc/rfc8707.html#name-access-token-request) does), it's clearer to define a new grant type specifically for requesting access tokens within an organization context.
+
+> The primary challenge of adding more parameters for this specification is the need to alter the inner OP implementation. Since we are creating an extension based on traditional specifications, there's no guarantee that upstream will accept our changes. Therefore, we opt to define a new grant type.
+
+#### 6.1.1 Token parameters
+
+To request an access token within an organization context, the client MUST send a request to the Token Endpoint with the following parameters:
+
+- `grant_type` (REQUIRED): The grant type. The value MUST be `urn:logto:grant-type:organization_token`.
```

Comment:
By that mean any third-party build-in OAuth modules or SDKs can not seamlessly integrate the organization feature in Logto?  Could cause extra effort for developers to understand or integrate with Logto. 

A couple of related questions:

Does this mean that the newly defined `urn:logto:grant-type:organization_token` grant type can only be used to request an organization-specific `access_token`? It cannot be mixed with OAuth `access_tokens`. Neither `resource` nor `resource scopes` are acceptable.

The new grant type shares the same `refresh_token` as the `refresh token grant type`. Also, the same rotate strategies. 

This might be the major challenging new concept in this design. Could have a further discussion and alignment. cc: @logto-io/eng

Reply:
true and not true. custom grants are supported in OAuth, and i belive modern SDKs can handle this easily. in addtion, we reuse the refresh token granted from the auth flow, so the effot won't be huge.

Reply:
> Does this mean that the newly defined urn:logto:grant-type:organization_token grant type can only be used to request an organization-specific access_token? It cannot be mixed with OAuth access_tokens. Neither resource nor resource scopes are acceptable.

yes, like it defines. what not in the spec won't be accepted

Reply:
> This might be the major challenging new concept in this design. Could have a further discussion and alignment.

Sorry for the confusion. The "new concept" refers to the new `Grant Type` here.

Reply:
discussed offline. keep as it is